app/Support/ItemCollection.php

```diff
@@ -492,10 +511,22 @@ public function canBlockLasers()
      *
      * @return bool
      */
-    public function canExtendMagic($bars = 2.0)
+    public function canExtendMagic($world = null, $bars = 2.0)
     {
-        return ($this->has('QuarterMagic') ? 4 : ($this->has('HalfMagic') ? 2 : 1))
-            * ($this->bottleCount() + 1) >= $bars;
+        $magicModifier = 1.0;
+        if ($world !== null)
+        {
+            $difficultyLevel = $world->config('rom.HardMode');
+            if ($difficultyLevel === 1) {
+                $magicModifier = 0.5;
+            }
+            else if ($difficultyLevel === 2) {
+                $magicModifier = 0.25;
+            }
+        }
+        $baseMagic = ($this->has('QuarterMagic') ? 4 : ($this->has('HalfMagic') ? 2 : 1))
+            * ($this->bottleCount() + 1);
+        return ($baseMagic * $magicModifier) >= $bars;
```

Comment:
$magicModifier should only apply to the bottle refill amount, not all magic usage.  i.e. `($this->bottleCount() * $magicModifier + 1)`
It does not apply in spike cave, since potions fully refill your magic there.
HardMode 3 still exists, and makes potions refill 0 magic.

Reply:
Thanks for the feedback.  I've changed the way it performs the calculation and added HardMode 3.  I also stopped passing in the $world object from the Spike Cave location so it won't apply the modifier.